app/modules/enduser_setup.c

```diff
@@ -478,90 +665,263 @@ static int enduser_setup_http_serve_html(struct espconn *http_client)
+
+    char *p = http + hdr_sz; /* start body where we know it will be */
+    /* p[0] will be clobbered when we print the header, so fill it in last */
+    ++p;
+    for (struct bss_info *wn = arg; wn; wn = wn->next.stqe_next)
     {
-      ENDUSER_SETUP_ERROR_VOID("enduser_setup_http_recvcb failed. Unknown error code.", ENDUSER_SETUP_ERR_UNKOWN_ERROR, ENDUSER_SETUP_ERR_NONFATAL);
+      if (wn != arg)
+        *p++ = ',';
+
+      const char entry_start[] = "{\"ssid\":\"";
+      strcpy (p, entry_start);
+      p += sizeof (entry_start) -1;
+
+      p = escape_ssid (p, wn->ssid);
+
+      const char entry_mid[] = "\",\"rssi\":";
+      strcpy (p, entry_mid);
+      p += sizeof (entry_mid) -1;
+
+      p += c_sprintf (p, "%d", wn->rssi);
+
+      *p++ = '}';
     }
+    *p++ = ']';
+
+    size_t body_sz = (p - http) - hdr_sz;
+    c_sprintf (http, header_fmt, body_sz);
+    http[hdr_sz] = '['; /* Rewrite the \0 with the correct start of body */
+
+    notify_scan_listeners (http, hdr_sz + body_sz);
+    c_free (http);
+    return;
+  }
+
+serve_500:
+  notify_scan_listeners (http_header_500, LITLEN(http_header_500));
+}
+
+/* ---- end WiFi AP scan support ------------------------------------------- */
+
+
+static void enduser_setup_http_recvcb(void *arg, char *data, unsigned short data_len)
+{
+  ENDUSER_SETUP_DEBUG(lua_getstate(), "enduser_setup_http_recvcb");
+  if (!state)
+  {
+    ENDUSER_SETUP_DEBUG(lua_getstate(), "ignoring received data while stopped");
+    return;
   }
-  else if (c_strncmp(data, "GET ", 4) == 0)
+
+  struct espconn *http_client = (struct espconn *) arg;
+  if (c_strncmp(data, "GET ", 4) == 0)
   {
-    /* Reject requests that probably aren't relevant to free up resources. */
-    if (c_strncmp(data, "GET / ", 6) != 0)
+    if (c_strncmp(data + 4, "/ ", 2) == 0)
     {
-      ENDUSER_SETUP_DEBUG(state->lua_L, "enduser_setup_http_recvcb received too specific request.");
-      enduser_setup_http_serve_header(http_client, (char *) http_header_404, sizeof(http_header_404));
-      enduser_setup_http_disconnect(http_client);
-      return;
+      if (enduser_setup_http_serve_html(http_client) != 0)
+        ENDUSER_SETUP_ERROR_VOID("http_recvcb failed. Unable to send HTML.", ENDUSER_SETUP_ERR_UNKOWN_ERROR, ENDUSER_SETUP_ERR_NONFATAL);
     }
+    else if (c_strncmp(data + 4, "/aplist ", 8) == 0)
+    {
+      scan_listener_t *l = os_malloc (sizeof (scan_listener_t));
+      if (!l)
+        ENDUSER_SETUP_ERROR_VOID("out of memory", ENDUSER_SETUP_ERR_OUT_OF_MEMORY, ENDUSER_SETUP_ERR_NONFATAL);
+
+      c_memcpy (l->remote_ip, http_client->proto.tcp->remote_ip, 4);
+      l->remote_port = http_client->proto.tcp->remote_port;
+
+      bool already = (state->scan_listeners != NULL);
+
+      l->next = state->scan_listeners;
+      state->scan_listeners = l;
 
-    retval = enduser_setup_http_serve_html(http_client);
-    if (retval != 0)
+      if (!already)
+      {
+        if (!wifi_station_scan (NULL, on_scan_done))
```

Comment:
This will fail if WiFi is not in Station (or Station+SoftAP) mode. I had a situation today where I entered into enduser setup with wifi mode=0. The AP was correctly turned on when start() was called, but nothing in code considered turning on the Station from within enduser_setup.